design/external-authorization-design.md

```diff
@@ -0,0 +1,670 @@
+Until Contour adds support for the v3 Envoy API, only the version 2 of the GRPC authorization protocol can be configured.
+
+Operators can bind an authorization service to a `HTTPProxy` using a new field in the `VirtualHost` struct.
+This field configures which `SupportService` resource to use, and allows the operator to set initial authorization policy.
+Authorization policy can also be set on a `Route`, so that application owners can pass metadata and disable authorization on specific routes.
+
+As discussed in [#2459][3], the Envoy external authorization HTTP filter is configured on the HTTP Connection Manager.
+This means that the scope of how Contour can configure authorization depends on the structure of the HTTP configuration.
+For HTTP, there is only one HTTP Connection Manager that contains all the virtual hosts.
+Configuring a single authorization service for all the virtual hosts might be acceptable for some organizations, but it doesn't really fix the Contour multi-tenancy model.
+On the other hand, Contour configures HTTPS with a separate HTTP Connection Manager for each virtual host, so different authorization servers can naturally be attached to each HTTPS virtual host.
+
+The result of this is that authorization servers are not supported on HTTP.
+While it is convenient from an implementation perspective, this policy is also consistent with Contour's security-first posture.
+Note that using the TLS fallback certificate (for non-SNI clients) has the same HTTP Connection Manager properties as HTTP, so authentication servers also cannot be configured on virtual hosts that share the fallback certificate.
+
+## Detailed Design
+
+### SupportService Changes
+
+The `SupportService` CRD has API version `projectcontour.io/v1alpha1`.
+
+There are a number of benefits to creating a CRD to represent a supporting service:
+
+- The CRD directly generates an Envoy [Cluster][4] that Contour can use for any purpose.
+  This means that with a single new API, Contour can add support for authorization, rate limiting and logging support services.
+- A CRD gives Contour a way to communicate the operational status of the support service, which was a [desired][9] [goal][7].
+- A CRD allows the team operating the support service and the team operating Contour to collaborate more loosely.
+
+```Go
+// SupportProtocolVersion is the version of the GRPC protocol used
+// to request support services.
+type SupportProtocolVersion string
+
+// SupportProtocolVersion2 requests the "v2" support protocol version.
+const SupportProtocolVersion2 SupportProtocolVersion = "v2"
+
+type SupportService struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   SupportServiceSpec   `json:"spec,omitempty"`
+	Status SupportServiceStatus `json:"status,omitempty"`
+}
+
+type SupportServiceStatus struct {
+	Conditions []Condition `json:"conditions"`
```

Comment:
How do users write to this field? If I build my own auth server to handle my auth mechanism, who manages this field? It's outside the scope of Contour since we may not own the implementation and can't understand what values to write.

Reply:
It's a Contour CRD, so Contour updates the status.

Reply:
To be clear from another conversation had via Zoom, this statement isn't true. The intent is that both Contour & 3rd party services should update this field.

Reply:
> To be clear from another conversation had via Zoom, this statement isn't true. The intent is that both Contour & 3rd party services should update this field.

Personally, I would be surprised if that happened 🤷